File: verify/claims.c
/* Recompute every published value in verify/claims.tsv in C, and check that
 * each claim phrase appears in the README.
 *
 * Build: cc -std=c99 -O2 -Wall -o claimc verify/claims.c -lm
 * Run:   ./claimc <repo root>
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAXCOL 16
#define MAXLINE 2048

static int problems = 0;
static void fail(const char *m) { printf("  FAIL: %s\n", m); problems++; }

static char *slurp(const char *path) {
    FILE *f = fopen(path, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END);
    long n = ftell(f);
    rewind(f);
    char *b = malloc((size_t)n + 1);
    fread(b, 1, (size_t)n, f);
    b[n] = '\0';
    fclose(f);
    return b;
}

static int find_col(char *hdr[], int n, const char *name) {
    for (int i = 0; i < n; i++)
        if (strcmp(hdr[i], name) == 0) return i;
    fprintf(stderr, "no column %s\n", name);
    exit(1);
}

static int split_tab(char *line, char *out[], int max) {
    int n = 0;
    out[n++] = line;
    while (*line) {
        if (*line == '\t') {
            *line = '\0';
            if (n < max) out[n++] = line + 1;
        }
        line++;
    }
    return n;
}

int main(int argc, char **argv) {
    const char *root = argc > 1 ? argv[1] : ".";
    char path[512];

    snprintf(path, sizeof path, "%s/README.md", root);
    char *readme = slurp(path);
    if (!readme) { fprintf(stderr, "cannot read %s\n", path); return 1; }

    snprintf(path, sizeof path, "%s/verify/claims.tsv", root);
    FILE *f = fopen(path, "r");
    if (!f) { fprintf(stderr, "cannot read %s\n", path); return 1; }

    char line[MAXLINE];
    char *cols[MAXCOL];
    fgets(line, sizeof line, f);
    line[strcspn(line, "\r\n")] = '\0';
    int ncols = split_tab(line, cols, MAXCOL);
    int ci = find_col(cols, ncols, "id");
    int cr = find_col(cols, ncols, "rule");
    int ca = find_col(cols, ncols, "a");
    int cb = find_col(cols, ncols, "b");
    int cp = find_col(cols, ncols, "published");
    int cph = find_col(cols, ncols, "phrase");

    int rows = 0;
    double worst = 0;
    while (fgets(line, sizeof line, f)) {
        line[strcspn(line, "\r\n")] = '\0';
        if (!*line) continue;
        char *fld[MAXCOL];
        split_tab(line, fld, MAXCOL);
        rows++;

        double a = atof(fld[ca]);
        double b_val = (strcmp(fld[cb], "-") == 0) ? 0 : atof(fld[cb]);
        double pub = atof(fld[cp]);
        const char *rule = fld[cr];
        double got;

        if (strcmp(rule, "copy") == 0) got = a;
        else if (strcmp(rule, "round0") == 0) got = round(a);
        else if (strcmp(rule, "round1") == 0) got = round(a * 10) / 10;
        else if (strcmp(rule, "round2") == 0) got = round(a * 100) / 100;
        else if (strcmp(rule, "ratio0") == 0) got = round(a / b_val);
        else if (strcmp(rule, "ratio1") == 0) got = round(a / b_val * 10) / 10;
        else { char m[256]; snprintf(m, sizeof m, "unknown rule %s", rule); fail(m); continue; }

        double err = fabs(got - pub);
        if (err > 1e-9) {
            char m[256];
            snprintf(m, sizeof m, "%s: published %.4f, recomputed %.4f", fld[ci], pub, got);
            fail(m);
        }
        if (err > worst) worst = err;

        if (!strstr(readme, fld[cph])) {
            char m[512];
            snprintf(m, sizeof m, "%s: phrase '%s' not found in README", fld[ci], fld[cph]);
            fail(m);
        }
    }
    fclose(f);
    free(readme);

    if (problems) {
        printf("C: %d problem(s)\n", problems);
        return 1;
    }
    printf("C: %d values reproduced, largest error %.1e, all phrases found in README\n", rows, worst);
    return 0;
}

File: verify/claims.mjs
import { readFileSync } from "fs";
import { join } from "path";

const root = process.argv[2] || ".";
const readme = readFileSync(join(root, "README.md"), "utf-8");
const lines = readFileSync(join(root, "verify", "claims.tsv"), "utf-8")
  .trim().split("\n");

const hdr = lines[0].split("\t");
const idx = (n) => hdr.indexOf(n);

let bad = 0;
let rows = 0;
for (let i = 1; i < lines.length; i++) {
  const f = lines[i].split("\t");
  if (!f[0]) continue;
  rows++;
  const a = parseFloat(f[idx("a")]);
  const b = f[idx("b")] === "-" ? null : parseFloat(f[idx("b")]);
  const pub = parseFloat(f[idx("published")]);
  const rule = f[idx("rule")];
  const id = f[idx("id")];
  const phrase = f[idx("phrase")];

  let got;
  if (rule === "copy") got = a;
  else if (rule === "round0") got = Math.round(a);
  else if (rule === "round1") got = Math.round(a * 10) / 10;
  else if (rule === "round2") got = Math.round(a * 100) / 100;
  else if (rule === "ratio0") got = Math.round(a / b);
  else if (rule === "ratio1") got = Math.round(a / b * 10) / 10;
  else { console.log(`  FAIL: unknown rule ${rule}`); bad++; continue; }

  if (Math.abs(got - pub) > 1e-9) {
    console.log(`  FAIL: ${id}: published ${pub}, recomputed ${got}`);
    bad++;
  }
  if (!readme.includes(phrase)) {
    console.log(`  FAIL: ${id}: phrase not in README`);
    bad++;
  }
}

if (bad) {
  console.log(`JS: ${bad} problem(s)`);
  process.exit(1);
}
console.log(`JS: ${rows} values reproduced, all phrases found in README`);
